packages/theme-ui/src/color-modes.js

```diff
@@ -5,11 +5,18 @@ import { useThemeUI } from './context'
 import { createColorStyles } from './custom-properties'
 
 const STORAGE_KEY = 'theme-ui-color-mode'
-const HAS_STORAGE = typeof Storage !== 'undefined'
+const hasStorage = () => {
```

Comment:
Thanks! I suspect it would be better to keep this as a constant rather than calling a function for every time this stores something in `localStorage`. Does a check for the `window` object not work in the environment you're seeing this in?

Reply:
Thank you for the quick response! Yep, it's not enough, because if you block cookies and site data (chrome://settings/content/cookies) then just an attempt of accessing localStorage will throw an exception (SecurityError).

I don't it as a big issue - even though it's a function call - it's indistinguishable in terms of performance comparing to window.localStorage. It takes the way more time to actually save the string inside the storage.

Reply:
Another way we can do it - wrap up window.localStorage into try/catch for both get/set methods. Then we don't need this hasStorage check at all.

Reply:
I like the idea of the whole thing being wrapped in a try catch -- it could still show a warning if localStorage fails for some reason, right?

Reply:
Yes, we can do something like this:

```
const storage = {
  get: init => {
    try {
      return window.localStorage.getItem(STORAGE_KEY) || init
    } catch (e) {
      console.warn(e)
    }
  },

  set: value => {
    try {
      window.localStorage.setItem(STORAGE_KEY, value)
    } catch (e) {
      console.warn(e)
    }
  },
}
```

This try/catch inside each method is kinda ugly but simple and probably won't get more complex.

Reply:
The warning message in the console will be:
_DOMException: Failed to read the 'localStorage' property from 'Window': Access is denied for this document._

or we can use something more custom like:
_localStorage is disabled and theming might not work as expected. Please check your Site Settings._

Reply:
Yeah, I like this because it would also fail silently if *anything* else happens to not work with localStorage

Reply:
Updated. I'm fine with the warning message duplication (React itself doest it a lot) or we can put it as a const (something line STORAGE_WARNING_MESSAGE)

Reply:
Probably better to add a real error as well, like this:

console.warn(
    'localStorage is disabled and theming might not work as expected.',
    'Please check your Site Settings.',
    e
  )

fyi @jxnblk

Reply:
The only thing localStorage is used for is keeping the color mode across sessions. Theming should still work if localStorage is disabled

Reply:
So, 'localStorage is disabled and color mode might not work as expected.' then?

Reply:
I think that sounds fine -- otherwise, this looks ready to go. Would be nice to have a test for this, but no worries if you don't have time

Reply:
I'll try to create a test. It might be harder than usual - Jest mocks a method and we can throw an exception as an implementation. However, in our case, we have to throw an error during property access and not a function call. I think [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) can solve this case. Let me try.

Reply:
I've added a test. Finally found a practical use-case for [Proxy](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy) 😃